pkg/proxy/apis/kubeproxyconfig/validation/validation.go

```diff
@@ -140,17 +141,24 @@ func validateKubeProxyConntrackConfiguration(config kubeproxyconfig.KubeProxyCon
 	return allErrs
 }
 
-func validateProxyMode(mode kubeproxyconfig.ProxyMode, fldPath *field.Path) field.ErrorList {
+func validateProxyMode(config *kubeproxyconfig.KubeProxyConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	switch mode {
+
+	ipvsFeatureGate := fmt.Sprintf("%s=true", features.SupportIPVSProxyMode)
+	if config.Mode == kubeproxyconfig.ProxyModeIPVS && !strings.Contains(config.FeatureGates, ipvsFeatureGate) {
```

Comment:
context: https://github.com/kubernetes/kubernetes/pull/56319#discussion_r152906722

>People can enable IPVS feature gateway, but it doesn't mean they always want to use IPVS proxy mode. 

Agreed. So I loosen the conditions than what I originally  modified was. I think the current condition (`if config.Mode == kubeproxyconfig.ProxyModeIPVS && !strings.Contains(config.FeatureGates, ipvsFeatureGate)`) is needed.

This means, if people want to use ipvs mode, they should enable the feature gate first, right? @m1093782566 @MrHohn

Reply:
Otherwise, `getProxyMode` will return wrong result.
```go
	if utilfeature.DefaultFeatureGate.Enabled(features.SupportIPVSProxyMode) {
		if proxyMode == proxyModeIPVS {
			return tryIPVSProxy(iptver, ipsetver, kcompat)
		} else {
			glog.Warningf("Can't use ipvs proxier, trying iptables proxier")
			return tryIPTablesProxy(iptver, kcompat)
		}
	}
```

Reply:
> This means, if people want to use ipvs mode, they should enable the feature gate first, right? 

That's true.

```go
	ipvsFeatureGate := fmt.Sprintf("%s=true", features.SupportIPVSProxyMode)
	if config.Mode == kubeproxyconfig.ProxyModeIPVS && !strings.Contains(config.FeatureGates, ipvsFeatureGate) {
		errMsg := fmt.Sprintf("must enable feature gate %s when specifying proxy mode %s ", ipvsFeatureGate, kubeproxyconfig.ProxyModeIPVS)
		allErrs = append(allErrs, field.Invalid(fldPath.Child("ProxyMode"), string(config.Mode), errMsg))
	}
```

It's not correct as well.

If people doesn't enable IPVS feature gateway, then IPVS is an unknown mode. Kube-proxy will firstly try iptables mode, then userspace mode...

Reply:
Feature gateway and kube-proxy mode are different thing.

Reply:
> Otherwise, getProxyMode will return wrong result.

See comments below.